docs/RFCS/online_schema_change.md

```diff
@@ -0,0 +1,166 @@
+leases](https://github.com/cockroachdb/cockroach/docs/RFCS/table_descriptor_lease.md)
+which describes the mechanism for asynchronously distributing
+modifications to table descriptors. This RFC is concerned with the
+actual steps of performing a schema change.
+
+# Detailed design
+
+Table descriptors will be enhanced to contain information about an
+on-going schema change operation:
+
+```proto
+message TableDescriptor {
+  ...
+
+  oneof mutation {
+    optional IndexDescriptor add_index;
+    optional IndexDescriptor drop_index;
+    optional ColumnDescriptor add_column;
+    optional ColumnDescriptor drop_column;
+  }
+}
+```
+
+Additionally, index and column descriptors will be enhanced to contain
+a state:
+
+```proto
+enum State {
+  DELETE_ONLY;
+  WRITE_ONLY;
+  PUBLIC;
+}
+```
+
+The `DELETE_ONLY` state specifies that `UPDATE` and `DELETE`
+operations should only remove keys from the index or for the
+column. The `WRITE_ONLY` state specifies that `INSERT`, `UPDATE` and
+`DELETE` operations should maintain the data for the index or column:
+adding or deleting as necessary. For both states, the column or index
+is not used for read operations. Lastly, the `PUBLIC` state allows the
+index or column to be used normally.
+
+The state for an index or column moves from `DELETE_ONLY` to
+`WRITE_ONLY` to `PUBLIC` or in reverse. It is invalid to jump from the
+`DELETE_ONLY` to `PUBLIC` state or vice versa. By moving the state in
+these discrete steps it can be shown that that a cluster will not
+create invalid data when two consecutive versions of the descriptor
+are in use concurrently. The table lease mechanism will be used to
+ensure that there are only two active versions of a descriptor in use
+in a cluster and the scheme change implementation will ensure that
+these two versions follow the state transition invariant.
+
+When an index is added to a table a potentially time consuming
+backfill operation needs to be performed to create the index
+data. This operation is performed when the index is in the
+`WRITE_ONLY` state and completed before moving the index to the
+`PUBLIC` state. Backfilling creates index entries for every row in the
+table. As such, it is infeasible for this backfilling to be performed
+transactionally: the transaction would create an unreasonable number
+of write intents and user transactions could cause the transaction to
+abort. Instead, backfill be performed as a series of small
+transactions that process a fraction of the table at a time:
+
+```go
+  startKey := MakeTableIndexKey(desc.ID, desc.Primary)
+  endKey := startKey.PrefixEnd()
+  for startKey != endKey {
+    var lastKey roachpb.Key
+    err := db.Txn(func(txn *Txn) error {
+      txn.SetPriority(VeryLowPriority)
+      scan, err := txn.Scan(startKey, endKey, 1000)
+      if err != nil {
+        return err
+      }
+      lastKey = getLastKeyOfFullRowInScan(scan)
+      b := makeIndexKeysBatch(scan)
```

Comment:
It's worth adding a checkpoint somewhere every once in a while. perhaps we should create a separate table for checkpoints.